### ISSUE TITLE
[re-opening] read_json_auto: duckdb.IOException: IO Error: Cannot open file "xxx": Too many open files

### ISSUE THREAD

**Issue by timon-schmelzer-gcx** [duckdb/duckdb#6588]
### What happens?

When reading about 1000 json files, I recieve the following error:

`read_json_auto: duckdb.IOException: IO Error: Cannot open file "xxx": Too many open files`

This issue has already been reported and a potential fix has been implemented ( https://github.com/duckdb/duckdb/issues/6249). Unfortunately, the issue is still available with duckdb 0.7.1 (and also the most recent pre-realese version of 0.7.2).

### To Reproduce

```python
import shutil
from pathlib import Path

import duckdb
import pandas as pd

if __name__ == "__main__":
    mode = "json"  # "json", "csv"

    data_dir = Path("data")
    if data_dir.exists():
        shutil.rmtree(data_dir)
    data_dir.mkdir(exist_ok=True)

    for i in range(1000):
        df = pd.DataFrame({"a": [1, 2, 3]})

        if mode == "json":
            df.to_json(data_dir / f"file_{i}.json", lines=True, orient="records")
        else:
            df.to_csv(data_dir / f"file_{i}.csv")

    data = duckdb.sql(f"SELECT COUNT(*) FROM read_{mode}_auto('{data_dir}/*.{mode}');")

    print(data.fetchone())
```

When setting mode="csv", the code works as expected.

### OS:

macOS

### DuckDB Version:

0.7.1

### DuckDB Client:

Python

### Full Name:

Timon Schmelzer

### Affiliation:

grandcentrix

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by nicosuave** [comment]
Can confirm; seeing the same on my end. Have been loading in batches in the meantime.

**Comment by lnkuiper** [comment]
I have tested this in the Python client, the CLI, and our `unittest` executable, and cannot reproduce this even with 2000 files. When adding print statements to debug, I can verify that all 2000 files are closed (MacOS Ventura 13.0, with an M1 chip).

I will discuss this with the one who implemented our CSV reader and see if he has a hunch about what is happening on your end. I would love to be able to reproduce and fix this.

**Comment by lnkuiper** [comment]
With some help, I was able to reproduce this! I'm currently investigating.

**Comment by lnkuiper** [comment]
I have pushed a fix in #6613 - for now the workaround is to set the `sample_size` parameter of `read_json_auto` to less than your open file handle limit, e.g., 10.

**Comment by timon-schmelzer-gcx** [comment]
Thanks for the fix!

However, the workaround does not work for me. From the example above, I just changed one line:

`data = duckdb.sql(f"SELECT COUNT(*) FROM read_{mode}_auto('{data_dir}/*.{mode}', sample_size=10);")`

**Comment by timon-schmelzer-gcx** [comment]
This also does not work for me:

`data = duckdb.sql(f"SELECT COUNT(*) FROM read_{mode}_auto('{data_dir}/*.{mode}', auto_detect=False, columns={{'a': BIGINT}});")`

**Comment by lnkuiper** [comment]
> This also does not work for me:
> 
> `data = duckdb.sql(f"SELECT COUNT(*) FROM read_{mode}_auto('{data_dir}/*.{mode}', auto_detect=False, columns={{'a': BIGINT}});")`

What is the error here? To me it seems like the column specification is wrong. It should be `columns={a: 'BIGINT'}`.

That being said, I suppose auto-detection should be disabled completely for this to work (in 0.7.1):
```sql
select count(*) from read_ndjson('__TEST_DIR__/input*.json', columns={a: 'BIGINT'});
```

Both your examples work with the fix in my PR.

**Comment by timon-schmelzer-gcx** [comment]
Thanks for your example, it works for me! Actually all combinations of single quotes seem to work for me, so `columns={'a': BIGINT}`, `columns={a: 'BIGINT'}`, `columns={'a': 'BIGINT'}` and `columns={a: BIGINT}`. 

I guess my main mistake was to use `read_json_auto` in combination with `auto_detect=False`, which does not make sense (obviously). Replacing it with `read_ndjson` or `read_json` works perfectly.

**Comment by timon-schmelzer-gcx** [comment]
Hey @lnkuiper , there seems to be another issue with closing files & the `read_json` functionality, this time when reading gzip files. Here is the code:

```python
import shutil
from pathlib import Path

import duckdb
import pandas as pd

if __name__ == "__main__":
    mode = "json"  # "json", "csv"

    data_dir = Path("data")
    if data_dir.exists():
        shutil.rmtree(data_dir)
    data_dir.mkdir(exist_ok=True)

    for i in range(500):
        df = pd.DataFrame({"a": list(range(3))})

        if mode == "json":
            df.to_json(
                data_dir / f"file_{i}.json.gz",
                lines=True,
                orient="records",
                compression="gzip",
            )
        else:
            df.to_csv(data_dir / f"file_{i}.csv.gz", index=False, compression="gzip")

    if mode == "json":
        data = duckdb.sql(
            """SELECT COUNT(*)
            FROM read_json('data/file_*.json.gz', columns={a: 'BIGINT'});
        """
        )
    else:
        data = duckdb.sql(
            """SELECT COUNT(*)
            FROM read_csv('data/file_*.csv.gz', columns={a: 'BIGINT'}, header=True);
        """
        )

    print(data.fetchone())
```

Error:
```
Traceback (most recent call last):
  File "/Users/timonschmelzer/Code/test/run.py", line 41, in <module>
    print(data.fetchone())
duckdb.IOException: IO Error: Cannot open file "data/file_344.json.gz": Too many open files
```

When chaning to csv files or decreasing the number of files, the code works as expected.

**Comment by lnkuiper** [comment]
@timon-schmelzer-gcx Thanks for the detailed bug report! I am keeping track of the JSON issues and will tackle a bunch of them at once

**Comment by lnkuiper** [comment]
Closed via #7478

The problem may still occur for compressed CSV files/parquet, as `CompressedFile` does not close properly when calling `Close()`.
